snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,421 @@
+
+* This should be on the order of a few minutes.
+
+### Reliability
+
+* Data corruption
+
+    * Though it is generally recommended to stop application writes before executing the snapshot command, we will not do this automatically for several reasons:
+
+        * GCE and Amazon can create snapshots while the application is running.
+
+        * Stopping application writes cannot be done from the master and varies by application, so doing so will introduce unnecessary complexity and permission issues in the code.
+
+        * Most file systems and server applications are (and should be) able to restore inconsistent snapshots the same way as a disk that underwent an unclean shutdown.
+
+    * The data consistency would be best-effort only: e.g., call fsfreeze prior the snapshot on filesystems that support it.
+
+    * There are several proposed solutions that would enable the users to specify the action to perform prior/after the
+    snapshots: e.g. use pod annotations.
+
+* Snapshot failure
+
+    * Case: Failure during external process, such as during API call or upload
+
+        * Log error, retry until success (indefinitely)
+
+    * Case: Failure within Kubernetes, such as controller restarts
+
+        *  If the master restarts in the middle of a snapshot operation, then the controller will find a snapshot request in pending state and should be able to successfully finish the operation.
+
+## Solution Overview
+
+There are a few uniqueness related to snapshots:
+
+* Both users and admins might create snapshots. Users should only get access to the snapshots belong to their namespaces. For this aspect, snapshot objects should be in user namespace. Admins might to choose expose the snapshots they created to some users who have access to those volumes.
+
+* After snapshots are taken, users might use them to create new volumes or restore the existing volumes back the time when the snapshot is taken.
+
+* There are use cases that data from snapshots taken from one namespace need to be accessible by users in another namespace.
+
+* For security purpose, if a snapshot object is created by a user, kubernetes should prevent other users duplicating this object in a different namespace if they happen to get the snapshot name.
+
+* There might be some existing snapshots taken by admins/users and they want to use those snapshots through kubernetes API interface.
+
+* **Create:**
+
+    1. The user creates a `VolumeSnapshot` referencing a persistent volume claim bound to a persistent volume
+
+    2. The controller fulfils the `VolumeSnapshot` by creating a snapshot using the volume plugins.
+
+    3. A new object `VolumeSnapshotData` is created to represent the actual snapshot binding the `VolumeSnapshot` with
+       the on-disk snapshot.
+
+* **List:**
+
+    1. The user is able to list all the `VolumeSnapshot` objects in the namespace
+
+* **Delete:**
+
+    1. The user deletes the `VolumeSnapshot`
```

Comment:
I would emphasize somewhere that we don't implement any ReclaimPolicy, i.e. all snapshots are deleted when user removes `VolumeSnapshot`.